gpii/node_modules/userListeners/src/pcsc.js

```diff
@@ -0,0 +1,675 @@
+    if (that.pcsclite) {
+        that.pcsclite.close();
+        that.pcsclite = null;
+        value = true;
+    }
+    that.active = false;
+    promise.resolve(value);
+    return promise;
+};
+
+/**
+ * Monitors a card reader. When a card is presented to the reader, read it and raise the onTokenArrive event when
+ * a token is found.
+ *
+ * @param that {Component} The gpii.userListeners.pcsc instance.
+ * @param reader {CardReader} The reader.
+ */
+gpii.userListeners.pcsc.monitorReader = function (that, reader) {
+    fluid.log("Got card reader: ", reader.name);
+
+    reader.on("status", function (status) {
+        var action = gpii.userListeners.pcsc.getCardAction(reader.state, status.state);
+        if (action === "presented") {
+            // Get the data from it.
+            gpii.userListeners.pcsc.readCard(that, reader, status.atr).then(function (tag) {
+                that.currentToken = tag;
+                that.events.onTokenArrive.fire(that, that.currentToken);
+            }, function (err) {
+                fluid.log(err);
+                that.events.onCardRejected.fire(that);
+            });
+        } else if (action === "removed") {
+            if (that.currentToken) {
+                that.events.onTokenRemove.fire(that, that.currentToken);
+                that.currentToken = null;
+            }
+        }
+    });
+
+    reader.on("end", function () {
+        fluid.log("Lost card reader: ", reader.name);
+    });
+    reader.on("error", function (e) {
+        // The device has gone, or is in an unknown state - restart the user listener.
+        fluid.log("Card reader error (" + reader.name + "): ", e.message);
+        that.failed(e);
+    });
+};
+
+/**
+ * Get the card action, based on the state change.
+ *
+ * @param oldState {number} Bit-mask of the old state.
+ * @param newState {number} Bit-mask of the new state.
+ * @return {string} Action of the card: "presented", "removed", or "unchanged".
+ */
+gpii.userListeners.pcsc.getCardAction = function (oldState, newState) {
+
+    // Get the states that have changed to on.
+    var newStates = (oldState ^ newState) & newState;
+    var action;
+    if (newStates & gpii.userListeners.pcsc.SCARD_STATE_EMPTY) {
+        // The "empty" bit has been set.
+        action = "removed";
+    } else if (newStates & gpii.userListeners.pcsc.SCARD_STATE_PRESENT) {
+        // The "present" bit has been set.
+        action = "presented";
+    } else {
+        action = "unchanged";
+    }
+    return action;
+};
+
+/**
+ * Detect a card based on the ATR (Answer to reset), returning the appropriate member of knownCards.
+ *
+ * (contact-less cards don't actually have an ATR; it's emulated by the reader)
+ *
+ * @param atr {Buffer} The card's ATR.
+ * @return {Object} Item from gpii.userListeners.pcsc.knownCards, or null if unrecognised.
+ */
+gpii.userListeners.pcsc.detectCard = function (atr) {
+    var card = null;
+    if (atr instanceof Buffer) {
```

Comment:
Perhaps a failure if it isn't, rather than a noop?

Reply:
yeh, that's better